src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs

```diff
@@ -162,160 +281,115 @@ impl FlowEventStore for FlowEventStoreInMem {
     }
 
     #[tracing::instrument(level = "debug", skip_all, fields(?flow_type))]
-    fn get_last_system_flow_of_type(&self, flow_type: SystemFlowType) -> Option<FlowID> {
+    fn get_last_system_flow_id_of_type(&self, flow_type: SystemFlowType) -> Option<FlowID> {
         let state = self.inner.as_state();
         let g = state.lock().unwrap();
         g.system_flows_by_type
             .get(&flow_type)
             .and_then(|flows| flows.last().copied())
     }
 
-    #[tracing::instrument(level = "debug", skip_all, fields(%dataset_id, ?flow_type))]
-    fn get_flows_by_dataset_of_type(
+    #[tracing::instrument(level = "debug", skip_all, fields(%dataset_id, ?filters, ?pagination))]
+    fn get_all_flow_ids_by_dataset(
         &self,
         dataset_id: &DatasetID,
-        flow_type: DatasetFlowType,
+        filters: DatasetFlowFilters,
+        pagination: FlowPaginationOpts,
     ) -> FlowIDStream {
-        let dataset_id = dataset_id.clone();
-
-        // TODO: This should be a buffered stream so we don't lock per record
-        Box::pin(async_stream::try_stream! {
-            let borrowed_key = BorrowedFlowKeyDataset::new(&dataset_id, flow_type);
-
-            let mut pos = {
-                let state = self.inner.as_state();
-                let g = state.lock().unwrap();
-                g.typed_flows_by_dataset.get(borrowed_key.as_trait()).map_or(0, Vec::len)
-            };
-
-            loop {
-                if pos == 0 {
-                    break;
-                }
-
-                pos -= 1;
-
-                let next = {
-                    let state = self.inner.as_state();
-                    let g = state.lock().unwrap();
-                    g.typed_flows_by_dataset
-                        .get(borrowed_key.as_trait())
-                        .and_then(|flows| flows.get(pos).copied())
-                };
+        let flow_ids_page: Option<Vec<_>> = {
```

Comment:
You could do `.collect().unwrap_or_default()` to get rid of `Option<>` which would also eliminate the match below.

Reply:
Nice